feat: add text selection and copy functionality to TUI

BurgessTG opened the pull request:
## Summary
- Adds click-and-drag text selection with visual highlighting
- Implements copy to clipboard functionality via OSC52
- Supports both single-line and multi-line selection

## Features
- **Mouse Selection**: Click and drag to select text
- **Visual Feedback**: Selected text is highlighted with system-appropriate colors
- **Multi-line Support**: Select across multiple lines seamlessly
- **Clean Copy**: Text is copied without ANSI escape codes
- **Keyboard Shortcuts**: `Ctrl+Shift+C` or `Ctrl+Y` to copy
- **Constrained Highlighting**: Highlight stays within content boundaries

## Implementation Details
- Uses OSC52 escape sequences for clipboard integration
- Handles centered content layout with proper coordinate mapping
- Strips ANSI codes for clean clipboard content
- Visual highlight uses muted system colors for better UX

## Test Plan
1. Run OpenCode TUI
2. Click and drag to select text (single or multi-line)
3. Press `Ctrl+Shift+C` or `Ctrl+Y` to copy
4. Paste in another application - text should be clean without formatting codes
5. Verify highlight doesn't spill outside content area

## Screenshots
The selection highlight uses system-appropriate muted colors and stays within content boundaries:
- Single-line selection works with proper visual feedback
- Multi-line selection maintains alignment and formatting
- Highlight color is subtle and doesn't interfere with readability

🤖 Generated with [opencode](https://opencode.ai)

BurgessTG committed:
feat: Add interactive scrollbar to chat interface

- Implement visual scrollbar with click-to-jump and drag-to-scroll functionality
- Use OpenCode's overlay system for clean rendering without ANSI corruption
- Add mouse event handling for scrollbar interactions
- Theme-integrated styling with proper visual feedback
- Non-intrusive design that preserves chat content formatting

🤖 Generated with [opencode](https://opencode.ai)

Co-Authored-By: opencode <noreply@opencode.ai>

BurgessTG committed:
fix: merge with dev branch scroll tracking changes

- Add lastScroll field to track mouse wheel events
- Add BUGGED_SCROLL_KEYS map to prevent key injection during scrolling
- Integrate scroll bug fix while preserving scrollbar functionality
- Resolves merge conflict with main dev branch

BurgessTG committed:
fix: complete merge with dev branch

- Add M and m to BUGGED_SCROLL_KEYS map
- Preserve all mouse event handling for scrollbar
- All conflicts resolved and tested

BurgessTG committed:
Merge branch 'dev' into feat/interactive-scrollbar

Resolved conflicts:
- Preserved scrollbar mouse event handling (MouseWheelMsg, MouseClickMsg, MouseReleaseMsg, MouseMotionMsg)
- Integrated lastScroll tracking from dev branch
- Maintained compatibility with both features

BurgessTG committed:
feat: add text selection and copy functionality to TUI

- Implement click-and-drag text selection with visual highlighting
- Add multi-line selection support with proper coordinate mapping
- Integrate OSC52 for clipboard copy (Ctrl+Shift+C / Ctrl+Y)
- Constrain highlight to content area to prevent spillover
- Use system-appropriate muted colors for selection highlight
- Extract clean text without ANSI codes for clipboard
- Handle centered content layout correctly

This feature allows users to select and copy text from the chat
interface, making it easier to extract code snippets, commands,
or other content from OpenCode responses.

🤖 Generated with [opencode](https://opencode.ai)

Co-Authored-By: opencode <noreply@opencode.ai>

adamdotdevin commented:
i'll try to pull this down early tomorrow, or monday at latest. curious though, mind posting a video?

BurgessTG commented:
> i'll try to pull this down early tomorrow, or monday at latest. curious though, mind posting a video?


https://github.com/user-attachments/assets/eb7ca92f-b896-49d6-a928-f1403a0a41e9

this is how its looking currently

adamdotdevin commented:
is it selecting the left border in that video?

BurgessTG commented:
yes, formatting is a little weird grabbing the borders, trying to see if i can fix it. can be better

BurgessTG committed:
fix: improve text extraction and cleaning for clipboard

- Add comprehensive text cleaning function to remove UI formatting
- Strip border characters (┃, │, |, ║, ┆, ┊) from copied text
- Intelligently format prose as paragraphs (join lines)
- Preserve code block and list formatting when detected
- Remove excessive whitespace and normalize indentation
- Fix cache type reference (MessageCache instead of cache.Cache)

This ensures copied text is clean and properly formatted for
pasting into other applications without UI artifacts.

🤖 Generated with [opencode](https://opencode.ai)

Co-Authored-By: opencode <noreply@opencode.ai>

BurgessTheGamer commented:
https://github.com/user-attachments/assets/68a74403-0338-4547-b20e-da4774186328 thats much better

BurgessTG commented:
https://github.com/user-attachments/assets/7e89f355-59b9-4fb2-9196-d672a7c0f1af

thats is much better

BurgessTG commented:
still a weird character bug that happens on highlight sometimes but functional!

BurgessTG committed:
fix: handle UTF-8 characters properly in text selection highlighting

- Convert string to runes before slicing to avoid breaking multi-byte characters
- Fixes diamond question mark artifacts (�) that appeared when selection boundaries split UTF-8 characters
- Ensures proper highlighting for all Unicode text including emojis and accented characters

🤖 Generated with [opencode](https://opencode.ai)

Co-Authored-By: opencode <noreply@opencode.ai>

BurgessTG commented:
> still a weird character bug that happens on highlight sometimes but functional!

fixed 

https://github.com/user-attachments/assets/7808b123-918a-4073-9902-a442d22a2b05

BurgessTG committed:
feat: add Aiken language server support for Cardano smart contracts

Adds automatic LSP integration for Aiken (.ak files) with:
- Auto-detection of .ak files
- Automatic installation via aikup if not present
- Full language server capabilities (diagnostics, completion, hover)
- Updated documentation with configuration examples

This enables intelligent code assistance for Cardano smart contract development.

🤖 Generated with [opencode](https://opencode.ai)

Co-Authored-By: opencode <noreply@opencode.ai>

BurgessTG committed:
Revert "feat: add Aiken language server support for Cardano smart contracts"

This reverts commit 39ae605951189e1405002237b48d53956e6a788a.

BurgessTG committed:
feat: Add height-limited text input with viewport scrolling

- Implement natural growth up to 8 lines, then fixed height with scrolling
- Add automatic cursor visibility tracking within viewport
- Preserve chat visibility when editing long text content
- Maintain all existing text editing capabilities

Technical implementation:
- Enhanced textarea component with scrollOffset and viewport behavior
- Modified SetHeight to support MaxHeight-based growth limiting
- Added ensureCursorVisible() for automatic scroll adjustment
- Updated View() method to render only visible lines within viewport

🤖 Generated with [opencode](https://opencode.ai)

Co-Authored-By: opencode <noreply@opencode.ai>

BurgessTG committed:
docs: Restore and update AGENTS.md with new workflow requirements

- Restore AGENTS.md documentation and opencode-dev-launcher.sh script
- Add height-limited text input checkpoint to completed features
- Document new git workflow requirements:
  * Always save to personal repository first
  * Ask before submitting PRs (no auto-submission)
  * Separate PRs for each new feature
- Update current development status section
- Establish clear workflow for future development

🤖 Generated with [opencode](https://opencode.ai)

Co-Authored-By: opencode <noreply@opencode.ai>

BurgessTG committed:
feat: Add mouse interaction to text input scrollbar

- Add click-to-jump functionality on scrollbar track
- Implement drag-to-scroll with smooth thumb movement
- Add mouse state management (isDragging, dragStartY, dragStartOffset)
- Handle MouseClickMsg, MouseMotionMsg, and MouseReleaseMsg events
- Calculate proportional positioning for accurate scroll control
- Preserve existing keyboard scrolling (Page Up/Down)

Technical implementation:
- Added shouldShowScrollbar() helper method
- Enhanced handleScrollbarClick() with thumb vs track detection
- Implemented handleScrollbarDrag() with delta calculation
- Mouse coordinates properly mapped to scroll positions
- Drag state properly reset on mouse release

BurgessTG committed:
feat: Fix mouse interaction and move scrollbar to border edge

Mouse Interaction Fixes:
- Add coordinate translation in editor component for mouse events
- Account for border (1px) and prompt (2 chars) offset
- Translate MouseClickMsg, MouseMotionMsg, and MouseReleaseMsg coordinates
- Enable proper mouse interaction with text input scrollbar

Scrollbar Positioning Improvements:
- Move scrollbar from textarea content to editor container border
- Add renderEditorScrollbar() method to editor component
- Position scrollbar on right border edge for cleaner appearance
- Remove scrollbar overlay from textarea View() method
- Add ScrollOffset() accessor method to textarea

Technical Details:
- Mouse coordinates now properly translated: X-3, Y-2 for textarea position
- Scrollbar rendered at container level with proper border positioning
- Maintains all existing scrollbar functionality (click, drag, proportional)
- Cleaner visual separation between content and scrollbar

BurgessTG committed:
fix: Implement proper editor-level scrollbar mouse handling

Mouse Interaction Fixes:
- Move scrollbar mouse handling from textarea to editor component level
- Add mouse state tracking to editorComponent (isDragging, dragStartY, dragStartOffset)
- Implement handleEditorScrollbarClick() for click-to-jump and drag initiation
- Implement handleEditorScrollbarDrag() for smooth drag scrolling
- Add setTextareaScrollOffset() to control textarea scroll position

Coordinate System Fixes:
- Handle mouse coordinates at editor level before translation
- Check scrollbar clicks at container coordinates (width-1, accounting for borders)
- Proper scrollbar area detection (Y: 2 to 2+visibleLines-1)
- Convert to scrollbar-relative coordinates for accurate positioning

Technical Implementation:
- Add clamp() utility method for bounds checking
- Preserve existing textarea mouse handling for non-scrollbar clicks
- Maintain drag state across mouse motion events
- Reset drag state on mouse release

This fixes the clicking issue by handling scrollbar interactions at the correct
coordinate system level (editor container) rather than translated textarea coordinates.

BurgessTG committed:
feat: Add clickable scrollbar to text input box

- Implemented scrollbar overlay for textarea when content exceeds MaxHeight
- Added ScrollOffset() and SetScrollOffset() methods to textarea
- Added mouse event handling for scrollbar clicks and dragging
- Styled scrollbar to match OpenCode design (thin track, solid thumb)
- Scrollbar appears inside the right border as an overlay

BurgessTG committed:
fix: Correct scrollbar positioning in text input

- Fixed scrollbar placement to account for border and padding
- Scrollbar now properly spans the content area (lines 2 to n-2)
- Added debug logging to help diagnose positioning issues
- Ensures scrollbar doesn't overlap with border corners

BurgessTG committed:
feat: Remove top/bottom borders with scrollbar and improve smoothness

- Remove top and bottom borders when scrollbar is present for cleaner look
- Adjust scrollbar positioning for borderless mode (padding only)
- Update mouse click/drag handlers for new positioning
- Prevent ensureCursorVisible from interfering with scrollbar control
- Scrollbar now provides smooth scrolling without cursor jumping

BurgessTG committed:
fix: Always remove top/bottom borders from text input

- Text input now never shows top/bottom borders, not just when scrollbar appears
- Gives consistent clean look matching OpenCode style
- Updated mouse coordinate adjustments for borderless design

BurgessTG committed:
debug: Add detailed logging for scrollbar interaction

- Log mouse click coordinates and scrollbar detection
- Log scrollbar state calculations (thumb position, size, etc)
- Log drag operations and position updates
- Track mouse release events for drag termination
- Will help diagnose clicking and scrolling issues

BurgessTG committed:
fix: Reduce render logging and improve click detection

- Remove frequent render logging that floods logs
- Keep only interaction logging (clicks, drags)
- Add scrollbarX to debug output to verify position
- This will help identify click detection issues

BurgessTG committed:
fix: Prevent cursor from jumping view back when using scrollbar

- Add manualScrolling flag to textarea to track scrollbar usage
- Skip ensureCursorVisible when manually scrolling
- Reset flag when user types or uses keyboard navigation
- Fixes issue where scrollbar would jump back to cursor position

BurgessTG committed:
refactor: Remove buggy scrollbar click/drag implementation

- Keep visual scrollbar rendering (works great)
- Remove problematic click/drag handling
- Simplify mouse event pass-through to textarea
- Ready for better interaction design

TODO: Implement better scrollbar interaction with:
- Keyboard shortcuts (Page Up/Down)
- Mouse wheel support
- Better hit detection
- Visual feedback

BurgessTG committed:
feat: Add mouse wheel support for textarea scrolling

- Scroll 3 lines per wheel tick
- Works when content exceeds MaxHeight
- Sets manualScrolling flag to prevent cursor jumping
- Uses Y field for direction (negative=up, positive=down)

Phase 1 of improved scrollbar interaction complete

BurgessTG committed:
fix: Route mouse wheel events to editor when hovering over it

- Add MouseWheelMsg to coordinate extraction in TUI
- Add MouseWheelMsg to editor bounds checking
- Pass MouseWheelMsg from editor to textarea
- Now mouse wheel scrolls textarea when hovering over it
- Main chat scrolls only when not over editor

BurgessTG committed:
fix: Use correct Button field for mouse wheel direction

- MouseWheelMsg uses Button field, not Y field for direction
- Y is the mouse position, not scroll direction
- Check for tea.MouseWheelUp and tea.MouseWheelDown buttons
- Add debug logging to show full message structure

BurgessTG committed:
feat: Add scrollbar click detection (Phase 1)

- Add ScrollbarState struct to track visual and interaction state
- Update scrollbar state on each render
- Detect clicks within scrollbar area (3-char tolerance)
- Distinguish between thumb and track clicks
- Add debug logging for click events

Ready for testing before implementing drag behavior

BurgessTG committed:
feat: Improve scrollbar drag handling

- Add dragStartScroll to track initial scroll position
- Add debug logging for mouse motion during drag
- Store scroll offset when drag starts for smoother calculation

Testing drag behavior to identify motion event issues

BurgessTG committed:
fix: Smooth scrollbar dragging and fix edge cases

- Use floating point math for smoother scroll calculations
- Fix drag range to allow scrolling to very bottom
- Simplify click-to-jump using percentage-based positioning
- Add bounds checking to prevent invalid positions
- Calculate scroll-per-pixel for consistent drag feel

Scrollbar should now drag smoothly from top to bottom

BurgessTG committed:
fix: Improve scrollbar precision and smoothness

- Track click position within thumb for natural dragging
- Use floating-point calculations for thumb size/position
- Rewrite drag handler to follow mouse precisely
- Ensure extremes (top/bottom) are always reachable
- Better ratio calculations for smooth movement

The scrollbar now:
- Follows mouse exactly where you clicked
- Moves smoothly without jumps
- Reaches top and bottom reliably
- Has proportional thumb size

BurgessTG committed:
fix: Improve scrollbar click detection reliability

- Add tolerance to hit detection boundaries (+1 pixel each side)
- Update scrollbar state before checking clicks
- Add debug logging for clicks near scrollbar
- Clamp click positions to valid range
- More forgiving edge detection

Should fix occasional missed clicks

BurgessTG committed:
feat: Hide cursor during scrollbar dragging in textarea

- Added SetScrollbarActive() method to textarea to control cursor visibility
- Hide cursor when starting scrollbar drag or track click
- Re-enable cursor when releasing mouse after drag
- Keep cursor visible during mouse wheel scrolling for better UX
- Prevents cursor interference with scrollbar interaction

BurgessTG committed:
feat: Move cursor to visible line after scrollbar interaction

- Added MoveCursorToVisibleLine() method to reposition cursor after scrolling
- Cursor now moves to the first visible line in viewport after scrollbar drag/click
- Prevents viewport from snapping back to original cursor position
- Provides smooth, expected behavior when using scrollbar

BurgessTG committed:
feat: Implement scrollLocked flag for better scrollbar behavior

- Added scrollLocked flag to prevent viewport snap-back after manual scrolling
- ensureCursorVisible now respects scrollLocked flag
- Improved MoveCursorToVisibleLine to position cursor in middle of viewport
- Added isCursorInViewport helper method
- On key press, if cursor is out of viewport while scrollLocked, move it to visible area
- Set scrollLocked after scrollbar drag/click and mouse wheel scrolling
- Removed aggressive cursor repositioning in favor of locked scroll state

This provides intuitive behavior where:
- Scrolling doesn't immediately move the cursor
- View stays where you scrolled until you type or navigate
- First keypress after scrolling moves cursor to viewport if needed
- Clicking in textarea will naturally position cursor at click location

BurgessTG committed:
feat: Add mouse click handling to textarea for cursor positioning

- Added tea.MouseClickMsg handling in textarea Update method
- Click positions cursor at the clicked location
- Handles wrapped lines correctly
- Accounts for line numbers if shown
- Clears scrollLocked flag on click for natural editing
- Added debug logging to help diagnose click issues

This fixes the issue where clicks in the textarea weren't registering
properly. Now clicking anywhere in the text positions the cursor there.

BurgessTG committed:
fix: Improve click detection accuracy in textarea

- Made scrollbar hit zone more precise (1 char wide instead of 4)
- Added debug logging for clicks passed to textarea
- This should reduce false positives where clicks near scrollbar
  were being intercepted when they shouldn't be

BurgessTG committed:
feat: Add hover effect to scrollbar for better visibility

- Track mouse motion to detect when hovering over scrollbar
- Change scrollbar appearance on hover:
  - Track changes from thin line │ to thick line ┃
  - Track color changes from BackgroundElement to Border
  - Thumb becomes bold on hover
- Trigger redraw when hover state changes
- Makes scrollbar more noticeable and easier to target

This improves UX by providing visual feedback when the mouse is
over the scrollbar, making it clearer where to click.

BurgessTG committed:
fix: Improve scrollbar hover detection and visual feedback

- Added separate isHoveringScrollbar method with wider hit zone (2 chars)
- Keep precise 1-char hit zone for actual clicks
- Enhanced hover visual feedback:
  - Track changes to thicker line with brighter color
  - Thumb gets background color and bold on hover
- Added debug logging for hover detection
- Simplified hover styling logic for better maintainability

This should make the hover effect more reliable and noticeable.

BurgessTG committed:
fix: Simplify editor scrollbar to match messages implementation

- Removed hover animation code completely
- Simplified track click to center thumb on click position (like messages)
- Simplified drag calculation to match messages scrollbar logic
- Removed complex offset calculations in favor of simpler approach
- This should make the scrollbar behavior more consistent and reliable

BurgessTG committed:
feat: Increase text input box height from 8 to 10 lines

- Provides more room for editing longer messages
- Better visibility of context when writing multi-line content
- May help with scrollbar interaction by providing more space

BurgessTG committed:
fix: Improve mouse click coordinate handling in textarea

- Added bounds checking to prevent negative coordinates
- Added detailed comments explaining the coordinate adjustments
- Added debug logging for negative coordinate cases
- Applied same fixes to mouse motion events
- This should fix missed clicks near the edges of the textarea

BurgessTG committed:
fix: Remove Y coordinate adjustment that was causing clicks to be one line off

- Removed the Y -= 1 adjustment that was causing clicks to register one line below
- The Y coordinate from the mouse event is already correct
- Updated debug logging to reflect the actual coordinates
- This should fix the click position being off by one line

BurgessTG committed:
fix: Adjust Y coordinate by -1 for top padding only

- Changed Y adjustment from -2 to -1
- The editor has top padding but no top border
- This should center clicks properly on the mouse position

BurgessTG committed:
fix: Remove Y coordinate adjustment entirely

- The Y coordinate from mouse events is already correct
- No adjustment needed for padding or borders
- This should fix clicks being one line below the mouse position

BurgessTG committed:
fix: Restore Y coordinate adjustment of -1 for top padding

- The textarea expects viewport-relative coordinates
- We need to adjust for the editor's top padding
- This should fix clicks being 2 lines below the mouse

BurgessTG committed:
fix: Adjust Y coordinate by -2 to fix one-line offset

- Changed Y adjustment from -1 to -2
- This accounts for top padding plus an additional line offset
- Should fix clicks being exactly one line off

BurgessTG committed:
fix: Widen scrollbar hit zone to 2 characters for easier clicking

- Accept clicks 1 character to the left of the scrollbar
- This gives a 2-character wide hit zone while keeping visual at 1 char
- Prevents accidental text selection when trying to click scrollbar
- Makes scrollbar easier to target with mouse

BurgessTG committed:
fix: Improve scrollbar top/bottom boundary detection

- Fixed bottom boundary check to use >= instead of >
- Added +1 tolerance to bottom boundary for easier clicking
- Consistent boundary checking in both isClickOnScrollbar and handleScrollbarClick
- This should fix clicks at the very top and bottom of scrollbar not registering

BurgessTG committed:
feat: Add interactive scrollbar to text input - 95% complete

- Implemented visual scrollbar with drag and click functionality
- Drag works flawlessly, click-to-jump mostly works
- Known issue: dynamic editor positioning causes Y-coordinate misalignment
- Updated AGENTS.md with current progress and technical challenges

BurgessTG committed:
feat: Interactive scrollbar 99% complete - only bottom edge issue remains

- Scrollbar fully functional except bottom 1-2 pixels
- Identified root cause: parent bounds check excludes bottom edge
- Updated AGENTS.md with detailed technical analysis

BurgessTG committed:
fix: Enable bottom edge clicks for editor scrollbar interaction

- Changed bounds check from < to <= for bottom edge detection
- Fixes issue where bottom 1-2 pixels of scrollbar were not clickable
- Scrollbar now fully interactive across entire height

delorenj commented:
> > still a weird character bug that happens on highlight sometimes but functional!
> 
> fixed
> 0628.3.mp4

This is 🔥 man!

BurgessTG committed:
fix: Update qr.go to use correct styles API

BurgessTG committed:
WIP: Resolving conflicts for PR 518

BurgessTG committed:
fix: Run go mod tidy to fix dependencies

BurgessTG committed:
Merge origin/dev into feat/text-selection-copy - resolved conflicts

likelymichael commented:
Why was this closed?

BurgessTG commented:
It doesn't work the same with all of the changes that have been made and also [image](https://github.com/user-attachments/assets/1551ab2f-4669-4d46-a3ba-551b28f628ad)

adamdotdevin commented:
hey @BurgessTheGamer, sorry for the lack of communication on my end. i would love for this feature to be a community contribution, but it's going to be tough logistically; this PR in particular is a pretty huge diff, and the first four files modified when i opened it up to review seem to be unrelated to the feature:
<img width="348" height="226" alt="Screenshot 2025-07-16 at 3 32 56 PM" src="https://github.com/user-attachments/assets/fae1aef9-7809-40f6-b01d-78ec7a5d49fd" />

it's hard to feel like a +2k line diff is going to be tight enough to merge with a start like that. happy to have a conversation if i misjudged this PR unfairly.

BurgessTG commented:
No worries, I'm still learning a handful of things, as well as this aspect of PRs. I spoke with my friend who is more experienced and he let me know something similar about the length, so no hard feelings. Just wanting to clear space 😄 this was definitely massive haha
